server/service/src/invoice/outbound_return/generate_lines/mod.rs

```diff
@@ -0,0 +1,384 @@
+
+        let return_lines = stock_line_repo
+            .query_by_filter(filter, Some(store_id.to_string()))?
+            .iter()
+            .map(stock_line_to_new_return_line)
+            .collect();
+
+        return_lines
+    } else {
+        vec![]
+    };
+
+    // If an item id is provided, we should return a new OutboundReturnLine for each stock line
+    // of that item that is currently available
+    let from_item_id = match &item_id {
+        Some(item_id) => {
+            let filter = StockLineFilter::new()
+                .item_id(EqualFilter::equal_to(item_id))
+                .is_available(true);
+
+            let return_lines = stock_line_repo
+                .query_by_filter(filter, Some(store_id.to_string()))?
+                .iter()
+                .map(stock_line_to_new_return_line)
+                .collect();
+
+            return_lines
+        }
+        None => vec![],
+    };
+
+    // If a return id is included alongside the item, we should include those existing return lines
+    // for that item
+    let existing_return_lines = match (item_id, return_id) {
+        (Some(item_id), Some(return_id)) => {
+            let lines = InvoiceLineRepository::new(&ctx.connection).query_by_filter(
+                InvoiceLineFilter::new()
+                    .invoice_id(EqualFilter::equal_to(&return_id))
+                    .item_id(EqualFilter::equal_to(&item_id)),
+            )?;
+
+            lines
+                .iter()
+                .map(|line| invoice_line_to_new_return_line(ctx, store_id, line))
+                .collect::<Result<Vec<OutboundReturnLine>, RepositoryError>>()?
+        }
+        _ => vec![],
+    };
+
+    // Filter out any stock lines that are already included in the existing return lines
+    let new_item_lines = from_item_id
+        .into_iter()
+        .filter(|new_line| {
+            !existing_return_lines.iter().any(|existing_line| {
+                new_line.stock_line.stock_line_row.id == existing_line.stock_line.stock_line_row.id
+            })
+        })
+        .collect::<Vec<OutboundReturnLine>>();
+
+    let mut return_lines = from_stock_line_ids;
+    return_lines.extend(existing_return_lines);
+    return_lines.extend(new_item_lines);
+
+    Ok(ListResult {
+        count: return_lines.len() as u32,
+        rows: return_lines,
+    })
+}
+
+fn stock_line_to_new_return_line(stock_line: &StockLine) -> OutboundReturnLine {
+    OutboundReturnLine {
+        id: uuid(),
+        stock_line: stock_line.clone(),
+        reason_id: None,
+        note: None,
+        number_of_packs: 0.0,
+    }
+}
+
+fn invoice_line_to_new_return_line(
+    ctx: &ServiceContext,
+    store_id: &str,
+    line: &InvoiceLine,
+) -> Result<OutboundReturnLine, RepositoryError> {
+    let stock_line_id = line
+        .invoice_line_row
+        .stock_line_id
+        .as_ref()
+        .ok_or(RepositoryError::NotFound)?;
```

Comment:
This might be a little bit lazy version... perhaps it would be better for me to create a specific service error here? Or try doing these checks in a validate method? Just noticed that most queries don't have any of that....